moto/cloudformation/parsing.py

```diff
@@ -343,8 +343,9 @@ def parse_and_update_resource(logical_id, resource_json, resources_map, region_n
         return None
 
 
-def parse_and_delete_resource(resource_name, resource_json, resources_map, region_name):
-    resource_class, resource_json, _ = parse_resource(resource_json, resources_map)
```

Comment:
The resources_map is a LazyDict, where resources are generated on the fly. If a resource is referenced that doesn't exist yet, it is automatically created. This lead to the workflow:
 - everything is created
 - call delete_stack
 - Delete the role
 - Start deleting the InstanceProfile (that references the role)
 - Parse the map to get all info from the InstanceProfile
 - Recreate the referenced role, as it no longer exist..
 - Finish deleting the InstanceProfile

Which lead to the fun bug where the role was deleted, but still existed afterwards.

Dict ordering ensured that it only occured in Python2. In Python3, it would delete the resources alphabetically; the InstanceProfile first, and the Role afterwards.

Reply:
Note that none of the implementations actually use this map, as far as I know, so the fact that it's now unparsed is irrelevant